### DISCUSSION TITLE
[Section 7.43] - Error: Incorrect attribute value type

### DISCUSSION THREAD

**Discussion by eduardocortez3** [chgasparoto/curso-aws-com-terraform#89] [Q&A]
Eu criei as subnets conforme sua demonstração na video-aula, utilizando o for_each. A minha dúvida é a seguinte, estou tentando pegar os subnet's id deste resource em um outro resource (elastic beanstalk), no campo settings deste resource (environment) o campo value só aceita valores string, tentei adicionar como você fez lá na video-aula do elastic load balancer sem sucesso, ele reclama o seguinte: 
Error: Incorrect attribute value type
│
│   on elastic_beanstalk.tf line 31, in resource "aws_elastic_beanstalk_environment" "this":
│   31:     value     = [aws_subnet.this["pub_a"].id, aws_subnet.this["pub_b"].id, aws_subnet.this["pub_c"].id]
│     ├────────────────
│     │ aws_subnet.this["pub_a"].id will be known only after apply
│     │ aws_subnet.this["pub_b"].id will be known only after apply
│     │ aws_subnet.this["pub_c"].id will be known only after apply
│
│ Inappropriate value for attribute "value": string required.

**Comment by chgasparoto** [comment]
@eduardocortez3 você está passando uma lista para a propriedade **value** que só aceita strings.
Isso talvez possa resolver o problema: `value = aws_subnet.this["pub_a"].id`

**Reply by eduardocortez3** [reply]
Entendi. Como eu poderia passar minhas 3 subnets públicas neste campo? 
No tutorial abaixo, o autor utiliza uma função chamada join, mas ele busca a variável em outro arquivo, essa variável é uma lista, o problema é que ele fixou as subnets id's. Veja parte do código.
`setting {
    namespace = "aws:ec2:vpc"
    name      = "Subnets"
    value     = join(",", var.public_subnets)
  }`
[https://automateinfra.com/2021/03/24/how-to-launch-aws-elastic-beanstalk-using-terraform/](url)

**Reply by chgasparoto** [reply]
@eduardocortez3 entendi. Você pode usar a função **join** ou concatenar as subnets.

Exemplo:
`value = join(",", [aws_subnet.this["pub_a"].id, aws_subnet.this["pub_b"].id, aws_subnet.this["pub_c"].id])`

ou algo do tipo

`value = "${aws_subnet.this["pub_a"].id}, ${aws_subnet.this["pub_b"].id}, ${aws_subnet.this["pub_c"].id}"`

**Reply by eduardocortez3** [reply]
Eu fiz da forma abaixo, funcionou também:
value     = "[ ${join(", ", [for k, v in aws_subnet.this : format("%q", v.id) if substr(v.tags.Name, 0, 3) == "Pub"])} ]"

Att.
